File: scala/67.scala
// JAVA_OPTS="-Xmx1024M" time scala -deprecation 67.scala
// 7273
// 1.71user 0.14system 0:03.66elapsed 50%CPU (0avgtext+0avgdata 0maxresident)k
// 8inputs+0outputs (0major+11788minor)pagefaults 0swaps

import scala.io.Source

val data = Source.fromFile("../data/67.txt")
                 .getLines
                 .toList
                 .map(_.trim.split(" ").map(_.toLong))
                 .toArray

for (i <- data.length - 2 to 0 by -1;
     j <- 0 until data(i).length) {
  val left = data(i + 1)(j)
  val right = data(i + 1)(j + 1)
  data(i)(j) += left max right
}

val result = data(0)(0)
println(result)

File: ruby/67.rb
data = []
file = File.new("../data/67.txt")
while line = file.gets
  data << line
end
file.close

data = data.map { |l| l.strip }.
            select { |l| l != "" }.
            map { |l| l.split(" ").map { |s| s.to_i } }

(data.length - 2).step(0, -1).each do |i|
  (0...data[i].length).each do |j|
    left = data[i+1][j]
    right = data[i+1][j+1]
    data[i][j] += [left, right].max
  end
end

puts data[0][0]
